src/write/str.rs

```diff
@@ -0,0 +1,182 @@
+use indexmap::IndexSet;
+use std::ops::{Deref, DerefMut};
+use vec::Vec;
+
+use common::DebugStrOffset;
+use write::{Result, Section, SectionId, Writer};
+
+/// An identifier for a string in a `StringTable.`
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct StringId(usize);
+
+/// A table of strings that will be stored in a `.debug_str` section.
+// Requirements:
+// - values are `[u8]`, null bytes are not allowed
+// - insertion returns a fixed id
+// - inserting a duplicate returns the id of the existing value
+// - able to convert an id to a section offset
+// Optional?
+// - able to get an existing value given an id
+//
+// Limitations of current implementation (using IndexSet):
+// - inserting a duplicate requires an allocation or a double lookup
```

Comment:
This is one of my longest standing annoyances with hash tables in Rust in general (mostly std, but apparently indexmap as well). I suspect a duplicated lookup is probably cheaper than an allocation (I haven't read further to see what you ended up implementing).

Reply:
It's actually allocation for duplicate, and double lookup for non-duplicate, so will depend on how often there are duplicates. I did the simplest (allocation), and we can change based on benchmarks.